### ISSUE TITLE
Making a type alias pub interpreted as making a trait sealed

### ISSUE THREAD

**Issue by kornelski** [obi1kenobi/cargo-semver-checks#1341]
### Which lint or lints are the issue

trait_newly_sealed

### Known issues that might be causing this

- [ ] Is the flagged item defined in another crate, or a re-export of such an item? ([#638](https://github.com/obi1kenobi/cargo-semver-checks/issues/638))

### Steps to reproduce the bug with the above code

baseline:

```rust
type TypeAlias = u32;

pub trait Foo {
    fn foo() -> TypeAlias;
}
```

current:

```rust
pub type TypeAlias = u32;

pub trait Foo {
    fn foo() -> TypeAlias;
}
```

### Actual Behaviour

```diff
@@ -8,11 +8,11 @@
             "crate_id": 0,
             "name": "TypeAlias",
             "span": {
-                "filename": "/private/tmp/baseline/src/lib.rs",
+                "filename": "/private/tmp/current/src/lib.rs",
                 "begin": [1, 1],
             },
-            "visibility": "crate",
+            "visibility": "public",
             "docs": null,
             "links": {},
             "attrs": [],
@@ -70,7 +70,16 @@
                     "sig": {
                         "inputs": [],
                         "output": {
-                            "primitive": "u32"
+                            "resolved_path": {
+                                "path": "TypeAlias",
+                                "id": 0,
+                                "args": {
+                                    "angle_bracketed": {
+                                        "args": [],
+                                        "constraints": []
+                                    }
+                                }
+                            }
                         },
                         "is_c_variadic": false
                     },
```

### Expected Behaviour

I think there are two separate issues here:

1. The `TypeAlias` doesn't seal. It's an alias of a public type. It has no generic arguments that could make it private. 
2. There's enough information in the JSON to resolve the `TypeAlias` to the aliased type, so it should be possible to see that the return type hasn't changed at all.

### Verbose Lint Output

> --- failure trait_newly_sealed: pub trait became sealed ---
>
> Description:
> A publicly-visible trait became sealed, so downstream crates are no longer able to implement it
>         ref: https://rust-lang.github.io/api-guidelines/future-proofing.html#sealed-traits-protect-against-downstream-implementations-c-sealed
>       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.41.0/src/lints/trait_newly_sealed.ron

### Generated System Information

(not relevant)

**Comment by obi1kenobi** [comment]
Thanks for reporting this. It's definitely a bug.

Type aliases are a mess in general. You've probably already seen #1254 which also mentions problems in the trait-sealing analysis caused by type aliases. I would be shocked if there aren't more of them too.

That said, these two statements aren't quite true, or at least are incomplete:

> The `TypeAlias` doesn't seal. It's an alias of a public type. It has no generic arguments that could make it private.

`TypeAlias` doesn't seal here. But in general, an alias of a public as-in-literally-`pub` type can seal in many ways. Here are two:
- The type could be `#[doc(hidden)]` i.e. `pub` but not public API. In this case, the trait _might be_ [sealed in its public API](https://predr.ag/blog/when-is-trait-can-be-implemented-public-api/) in this case, even though it wouldn't be unconditionally sealed.
- The type might not be constructible with a literal, and might have no functions that return it nor any `pub` fields of that type to destructure in its crate. `transmute()` is not generally considered to be a "public API" way to make a type, so it doesn't count.

> There's enough information in the JSON to resolve the `TypeAlias` to the aliased type, so it should be possible to see that the return type hasn't changed at all.

Outside of the case of primitives like `u32`, this is true only if we assume we implicitly inline `std` and `core` into the rustdoc JSON here. This isn't feasible today, for a variety of unfortunate but not fundamental reasons.

Otherwise: the original return type may have used one path like `std::Foo`, the type alias could be using another path like `core::Foo`, and they might happen to be equivalent. But we wouldn't have a way to know that.

***

Are type aliases critical to your use case? Assuming we're able to agree on a substantial ongoing sponsorship, I'd be happy to learn more about how you use them and do my best to make those code paths work as best as possible.

I just want to be honest and upfront here: a complete end-to-end fix of everything wrong with type aliases (including issues like #1254) will take years of focused work. They are very complicated! I'm hoping we can somewhat quickly 80-20 this problem for your use cases specifically, while we also (slowly) make progress on the long tail of issues.

**Comment by kornelski** [comment]
In my specific case, I'm cleaning up `type BoxError = Box<dyn std::error::Error>` that has been redundantly defined privately in a few places, and replacing it with `crate::DynError` alias of the same type, that happens to be `pub`. It has no effect on the public API, because all the aliases type check as `Box<dyn std::error::Error>`, and nothing public has been removed. It's not a blocker for me. I can simply ignore the false negative and merge the code anyway.

Changing a type alias *from* `pub` to private or hidden definitely has a potential to seal, but I don't see how going the other way, from private *to* `pub, can ever\* seal. Making a previously inaccessible type alias accessible only gives access to more types. Are you sure we're not talking past each other due to misunderstanding of the direction of the change?

----

\* I'm not counting the edge case of new public items shadowing glob imports of different items with the same name, which is a way to make things private by adding a new `pub` item. That's a general issue with Rust's imports, not specific to type aliases or sealing. You don't need to handle this explicitly, because `rustdoc` resolves imports for you, so the shadowed item would disappear from the JSON if that happened, and you'd see this directly.

**Comment by wallstop** [comment]
Verified against 0.50.0 (trustfall-rustdoc-adapter 57.1.0, rustdoc JSON format 57): the same alias-signature sealing analysis also suppresses **`trait_method_added`** — adding a *required* method to a public trait is a silent false negative when any required method's signature mentions a crate-defined type alias. This direction matters for release gating: the lint's baseline filter (`sealed_trait != true`) drops the new-method finding entirely, so an unconditionally-implementable trait gains a required method without any semver failure.

Minimal repro (verified locally):

baseline:
```rust
pub type Result<T> = std::result::Result<T, ()>;

pub trait Api {
    fn a(&self) -> Result<()>;
}
```

current:
```rust
pub type Result<T> = std::result::Result<T, ()>;

pub trait Api {
    fn a(&self) -> Result<()>;
    fn b(&self) -> u32; // new required method — never reported
}
```

Result: 223 checks pass, `trait_method_added` skips the trait. Control: changing `a` to return `u32` in both revisions makes `trait_method_added` fire as expected, so the alias in the *sibling* method's signature is what taints the sealing analysis for the whole trait.

We hit this in production when `transport_diagnostics` was added as a required method to our public client trait (every method returns the crate's `error::Result<T>` alias); the changelog's breaking marker was the only working gate. Real-world reduced repro: https://github.com/Ambiguous-Interactive/signal-fish-client-rust/blob/main/src/client_api.rs
